packages/playground-preview-worker/src/user.do.ts

```diff
@@ -0,0 +1,232 @@
+			}
+			this.workerName = workerName;
+
+			this.previewToken = await this.state.storage.get<string>("previewToken");
+			this.inspectorUrl = await this.state.storage.get<string>("inspectorUrl");
+		});
+	}
+	async refreshTokens() {
+		this.config = await setupTokens(this.env.ACCOUNT_ID, this.env.API_TOKEN);
+		await this.state.storage.put("config", this.config);
+	}
+	async uploadWorker(name: string, worker: FormData) {
+		worker.set(
+			"wrangler-session-config",
+			JSON.stringify({ workers_dev: true })
+		);
+		if (this.config === undefined) {
+			console.log("Setting up tokens");
+			await this.refreshTokens();
+			assert(this.config !== undefined);
+		}
+		let uploadResult: UploadResult;
+		try {
+			uploadResult = await doUpload(
+				this.env.ACCOUNT_ID,
+				this.env.API_TOKEN,
+				this.config,
+				name,
+				worker
+			);
+		} catch (e) {
+			// Try to recover _once_ from failure. This captures expired tokens, but means that genuine failures won't cause
+			// a request loop and will return an error to the user
+			await this.refreshTokens();
+			uploadResult = await doUpload(
+				this.env.ACCOUNT_ID,
+				this.env.API_TOKEN,
+				this.config,
+				name,
+				worker
+			);
+		}
+
+		const inspector = new URL(
+			this.config.uploadConfigToken.inspector_websocket
+		);
+		inspector.searchParams.append(
+			"cf_workers_preview_token",
+			this.config.uploadConfigToken.token
+		);
+		inspector.protocol = "https:";
+		// Fire and forget
+		void fetch(this.config.uploadConfigToken.prewarm, {
+			method: "POST",
+			headers: {
+				"cf-workers-preview-token": uploadResult.result.preview_token,
+			},
+		});
+		this.previewToken = uploadResult.result.preview_token;
+		this.inspectorUrl = inspector.href;
+
+		await this.state.storage.put("previewToken", this.previewToken);
+		await this.state.storage.put("inspectorUrl", this.inspectorUrl);
+	}
+
+	async fetch(request: Request) {
+		const url = new URL(request.url);
+		// We need to construct a new Sentry instance here because throwing
+		// errors across a DO boundary will wipe stack information etc...
+		const sentry = setupSentry(
+			request,
+			undefined,
+			this.env.SENTRY_DSN,
+			this.env.SENTRY_ACCESS_CLIENT_ID,
+			this.env.SENTRY_ACCESS_CLIENT_SECRET
+		);
+		// This is an inspector request. Forward to the correct inspector URL
+		if (request.headers.get("Upgrade") && url.pathname === "/api/inspector") {
+			assert(this.inspectorUrl !== undefined);
+			return fetch(this.inspectorUrl, request);
+		}
+		// This is a request to run the user-worker. Forward, adding the correct authentication headers
+		if (request.headers.has("cf-run-user-worker")) {
+			assert(this.previewToken !== undefined);
+
+			const workerResponse = await fetch(
+				switchRemote(
+					new URL(request.url),
+					`https://${this.workerName}.${this.env.workersDev}`
+				),
+				new Request(request, {
+					headers: {
+						...Object.fromEntries(request.headers),
```

Comment:
we need to be sure to remove `x-cf-token` so we don't accidentally send it twice and blow request size limits. See https://github.com/cloudflare/workers-sdk/pull/3722